add ssupport for a few more opinfos in jit

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59814 Test shape analysis with opinfos
* **#59812 add ssupport for a few more opinfos in jit**
* #59811 Don't substitute in symbolic shapes to shape compute graph
* #59810 small cleanups
* #59809 Only optimize after change (redo)

This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

Differential Revision: [D30200061](https://our.internmc.facebook.com/intern/diff/D30200061)

Commit:
add ssupport for a few more opinfos in jit

[ghstack-poisoned]

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -397,6 +397,15 @@ def test_variant_consistency_jit(self, device, dtype, op):
+                # lambdas are typically used as a way to simulate methods without
+                # functional variants, so rely on the other variant for testing
+                # for now
+                if isinstance(variant, LambdaType):
```
Should we just use this approach instead of adding metadata to the OpInfo?

eellison replied:
has_fake_function != is_lambda

mruberry commented:
This is cool but looks like there are still a lot of test failures.

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

[ghstack-poisoned]

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -771,20 +777,37 @@ def test_variant_consistency_jit(self, device, dtype, op):
+        # TODO: find better way to standardize on op registration itself..
+        has_fake_function = op.name in ["resize_", 'resize_as_']
+
+        if has_fake_function:
```
I would just add this attribute to the OpInfos for now

eellison replied:
fine doing that, for two functions it seems like a bit of overkill... but i dont know how prevalent it will be elsewhere

mruberry commented on test/test_ops.py:
```diff
@@ -771,20 +777,37 @@ def test_variant_consistency_jit(self, device, dtype, op):
+                # lambdas are typically used as a way to simulate methods without
+                # functional variants, so rely on the other variant for testing
+                # for now
+                if is_lambda(variant):
```
Would this be replaced by querying the "has_fake_function" attribute?

eellison replied:
is `has_fake_function` exactly the same thing as lambda ? I think the fake function is where we are adding extra nodes as a way to not mutate the inputs. There are other nodes with a lambda like torch.where

mruberry replied:
It can be whatever we want. I'd design it so it's most useful to you so long as it's clear to users how to set the metadata

mruberry commented on test/test_ops.py:
```diff
@@ -771,20 +777,37 @@ def test_variant_consistency_jit(self, device, dtype, op):
+                tested = True
+
                 # Create accessor for script function variant
                 name = op.name + '_' if func_type == 'inplace' else op.name
```
I think this can be removed because `func_type` can never be "inplace" with the current filtering

eellison replied:
will add assertion instead

mruberry commented on test/test_ops.py:
```diff
@@ -801,30 +824,35 @@ def out_fn(output):
                         return output
 
+                    def get_sample():
+                        return clone_input_helper(sample.input) if op.name[-1] == '_' else sample.input
```
I don't think this helper is needed if the variant can never be inplace

eellison replied:
hmm do we know that ? i guess i'll add an assertion

eellison reviewed (commented):
@mruberry i was having some issues across the two separate PRs so I merged them into one. sorry for the churn!

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

[ghstack-poisoned]

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200061).

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

Differential Revision: [D30200061](https://our.internmc.facebook.com/intern/diff/D30200061)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200061).

Commit:
Update on "add ssupport for a few more opinfos in jit"


This is sort of a half measure: we can successfully trace through opinfos which are registered as lambdas, we just can't script them. This tests if the op is a lambda in which case bails... see the next PR to get resize_ to work, maybe this should be consolidated with that...

Differential Revision: [D30200061](https://our.internmc.facebook.com/intern/diff/D30200061)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200061).
